boa/src/builtins/array/array_iterator.rs

```diff
@@ -0,0 +1,149 @@
+use crate::{
+    builtins::{
+        function::{make_builtin_fn, BuiltInFunction, Function, FunctionFlags},
+        Array, Value,
+    },
+    object::{Object, ObjectData, PROTOTYPE},
+    property::Property,
+    Context, Result,
+};
+use gc::{Finalize, Trace};
+use std::borrow::Borrow;
+
+#[derive(Debug, Clone, Finalize, Trace)]
+pub enum ArrayIterationKind {
+    Key,
+    Value,
+    KeyAndValue,
+}
+
+#[derive(Debug, Clone, Finalize, Trace)]
+pub struct ArrayIterator {
+    array: Value,
+    next_index: i32,
+    kind: ArrayIterationKind,
+}
+
+impl ArrayIterator {
+    fn new(array: Value, kind: ArrayIterationKind) -> Self {
+        ArrayIterator {
+            array,
+            kind,
+            next_index: 0,
+        }
+    }
+
+    pub(crate) fn new_array_iterator(
+        ctx: &Context,
+        array: Value,
+        kind: ArrayIterationKind,
+    ) -> Result<Value> {
+        let array_iterator = Value::new_object(Some(
+            &ctx.realm()
+                .environment
+                .get_global_object()
+                .expect("Could not get global object"),
+        ));
+        array_iterator.set_data(ObjectData::ArrayIterator(Self::new(array, kind)));
+        array_iterator
+            .as_object_mut()
+            .expect("array iterator object")
+            .set_prototype_instance(
+                ctx.realm()
+                    .environment
+                    .get_binding_value("Object")
+                    .expect("Object was not initialized")
+                    .borrow()
+                    .get_field(PROTOTYPE),
+            );
+        make_builtin_fn(Self::next, "next", &array_iterator, 0, ctx);
+        let mut function = Object::function(
+            Function::BuiltIn(
+                BuiltInFunction(|v, _, _| Ok(v.clone())),
+                FunctionFlags::CALLABLE,
+            ),
+            ctx.global_object()
+                .get_field("Function")
+                .get_field("prototype"),
+        );
```

Comment:
shouldn't this point to array iterator `prototype` and not `Object.prototype`
Also we set it to `Object.prototype` here:
```rust
let array_iterator = Value::new_object(Some(
            &ctx.realm()
                .environment
                .get_global_object()
                .expect("Could not get global object"),
        ));
```
and again here:
```rust
array_iterator
            .as_object_mut()
            .expect("array iterator object")
            .set_prototype_instance(
                ctx.realm()
                    .environment
                    .get_binding_value("Object")
                    .expect("Object was not initialized")
                    .borrow()
                    .get_field(PROTOTYPE),
            );
```


This should be in `<Array Iterator>.prototype.next`. This should only be done in the builtin initialization phase and access that prototype  when we need it. we can store it in `Context` for now and initialize it and use it where we need it.
```rust
make_builtin_fn(Self::next, "next", &array_iterator, 0, ctx);
```

Reply:
Could you clarify the implementation of this? Do you mean having an init function that creates a prototype instance that is then set as the prototype instance when an Array Iterator object is created?

Reply:
I tihnk I've figured it out. Would it be as a cache of iterator prototypes held in context, similar to the cache of well known symbols?

Reply:
> I think I've figured it out. Would it be as a cache of iterator prototypes held in context, similar to the cache of well known symbols?

Sorry for not explaning it that much, but its pretty much the same we would initialize it in the builtins initialization phase. and use it when we construct a Array iterator

Reply:
I've implemented something, would you be able to tell me if it's the right sort of thing?

Reply:
> I've implemented something, would you be able to tell me if it's the right sort of thing?

Yes. This looks very good! Definitely in the right direction :)